### PR TITLE
add NekoVM

### PR DESCRIPTION
**Pull request by antonl05**

### CODE REVIEW

File: packages/NekoVM/build.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+TERMUX_PKG_SRCURL=https://github.com/HaxeFoundation/neko/archive/refs/tags/v${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=850e7e317bdaf24ed652efeff89c1cb21380ca19f20e68a296c84f6bad4ee995
+TERMUX_PKG_ESSENTIAL=false
+TERMUX_PKG_DEPENDS="libgc,openssl,zlib,apache2,libsqlite,mbedtls"
+
+termux_step_configure() {
+        #moving files from src to build since build system looks inside build.
+        #building with ninja fails as it keeps looping.
+        cp -r ../src/* .
+        cmake -Wno-dev -DCMAKE_INSTALL_PREFIX:PATH=$TERMUX_PREFIX -DWITH_UI=n -DWITH_MYSQL=n -DWITH_REGEXP=n
```
Comment (xeffyr) [inline packages/NekoVM/build.sh:15]: You can't use cmake without doing `termux_setup_cmake` first. Also check the proper configuration options: https://github.com/termux/termux-packages/blob/master/scripts/build/configure/termux_step_configure_cmake.sh

You are cross-compiling.

File: packages/NekoVM/build.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+termux_step_make() {
+        make -j $(nproc)
```
Comment (xeffyr) [inline packages/NekoVM/build.sh:19]: Use `$TERMUX_MAKE_PROCESSES` instead of `$(nproc)`.

File: packages/NekoVM/build.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
+TERMUX_PKG_HOMEPAGE="https://nekovm.org/"
+TERMUX_PKG_DESCRIPTION="The Neko Virtual Machine"
+TERMUX_PKG_LICENSE="MIT"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=2-3-0
+TERMUX_PKG_SRCURL=https://github.com/HaxeFoundation/neko/archive/refs/tags/v${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=850e7e317bdaf24ed652efeff89c1cb21380ca19f20e68a296c84f6bad4ee995
+TERMUX_PKG_ESSENTIAL=false
```
Comment (xeffyr) [inline packages/NekoVM/build.sh:8]: Normally you shouldn't specify TERMUX_PKG_ESSENTIAL at all.

File: packages/NekoVM/build.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+TERMUX_PKG_HOMEPAGE="https://nekovm.org/"
+TERMUX_PKG_DESCRIPTION="The Neko Virtual Machine"
+TERMUX_PKG_LICENSE="MIT"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=2-3-0
```
Comment (xeffyr) [inline packages/NekoVM/build.sh:5]: This value cannot be used as version. Specify `2.3.0` there and do conversion in `TERMUX_PKG_SRCURL`.

**Review by xeffyr — CHANGES_REQUESTED** [review]

**Review by xeffyr — CHANGES_REQUESTED** [review]
Package name must be lowercase. Check review comments regarding other issues.

**Comment by antonl05** [comment]
please take a look at the releases page of nekoVM , the packages that they
uploaded are in the format of - rather than in _ , so until there is
another place to download the source archive , it cant be changed . . or
hardcode the version , which isnt hard to update in future

,or shall i create a repo for this one source package , so that i can
change this from - to _ , make another commit changing it in build.sh

please reply on the choice that you made as i am a bit nervous since this
is my first commit

On Jul 29, 2021 17:04, "Leonid Pliushch" ***@***.***> wrote:

***@***.**** requested changes on this pull request.

Package name must be lowercase. Check review comments regarding other
issues.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/termux/termux-packages/pull/7206#pullrequestreview-717984805>,
or unsubscribe

**Comment by xeffyr** [comment]
**Rename package**: `packages/NekoVM/build.sh` --> `packages/nekovm/build.sh`. CamelCase package names aren't accepted. Make it lowercase.

**Comment by xeffyr** [comment]
Ok, now fix cmake issue on which I have pointed in review comment (`termux_setup_cmake` + flags).
```
Downloading https://github.com/HaxeFoundation/neko/archive/refs/tags/v2-3-0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   137  100   137    0     0   1505      0 --:--:-- --:--:-- --:--:--  1505

100  368k    0  368k    0     0  1304k      0 --:--:-- --:--:-- --:--:-- 1304k
100  501k    0  501k    0     0  1662k      0 --:--:-- --:--:-- --:--:-- 6699k
/home/builder/termux-packages/packages/nekovm/build.sh: line 14: cmake: command not found
```

**Comment by antonl05** [comment]
seems that build system cant find gc/gc.h , which is supposed to be in system.
even though i compiles under my phone.

**Comment by xeffyr** [comment]
> even though i compiles under my phone.

Our build system is not Termux on phone. That's Ubuntu 20.04 Docker image with cross-compiler. So please read my previous comments apply necessary changes. Specifically follow https://github.com/termux/termux-packages/blob/master/scripts/build/configure/termux_step_configure_cmake.sh, if you decided to use own CMake command instead of predefined one.

File: packages/nekovm/build.sh [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+TERMUX_PKG_LICENSE="MIT"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=2.3.0
+TERMUX_PKG_SRCURL=https://github.com/HaxeFoundation/neko/archive/refs/tags/v${TERMUX_PKG_VERSION//[.]/-}.tar.gz
+TERMUX_PKG_SHA256=850e7e317bdaf24ed652efeff89c1cb21380ca19f20e68a296c84f6bad4ee995
+TERMUX_PKG_DEPENDS="libgc,openssl,zlib,apache2,libsqlite,mbedtls"
+TERMUX_PKG_SRCDIR="../src"
+
+
+termux_step_configure_cmake() {
```
Comment (xeffyr) [inline packages/nekovm/build.sh:12]: ???

You shouldn't override this step.

Read https://github.com/termux/termux-packages/wiki/Building-packages and also you have lots of sample packages in https://github.com/termux/termux-packages/tree/master/packages (including cmake-based).

File: packages/nekovm/build.sh [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                -DCMAKE_MAKE_PROGRAM=$MAKE_PROGRAM_PATH \
+                -DCMAKE_SKIP_INSTALL_RPATH=ON \
+                -DCMAKE_USE_SYSTEM_LIBRARIES=True \
+                "${CMAKE_ADDITIONAL_ARGS[@]}" \
+                -Wno-dev \
+                -DWITH_UI=OFF \
+                -DWITH_MYSQL=OFF \
+                -DWITH_REGEXP=OFF \
+                -DRUN_LDCONFIG=OFF \
+                $TERMUX_PKG_EXTRA_CONFIGURE_ARGS
```
Comment (xeffyr) [inline packages/nekovm/build.sh:65]: I see you have just copy-pasted the whole https://github.com/termux/termux-packages/blob/master/scripts/build/configure/termux_step_configure_cmake.sh without understanding what it does.

**Review by xeffyr — CHANGES_REQUESTED** [review]
Fix build.sh since it is completely messed up.

File: packages/nekovm/build.sh (hunk 6 above)
Comment (antonl05) [reply packages/nekovm/build.sh:65]: i was just follwing the standard

**Review by xeffyr — COMMENTED** [review]
I've did all the changes that had to be done by you.

*Is it hard to follow review recommendations nowadays?*

Resources for read:
* https://github.com/termux/termux-packages/wiki, whole, all pages, especially about packaging.
* https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md

*Maintainers won't do all your work for pull requests on their own. If you can't submit a good quality pull request, then it is better to just request the package on issue tracker.*

**Comment by xeffyr** [comment]
It now builds but it seems a host build is required. It tries to execute cross-compiled binary which is not possible.
```
/bin/sh: 1: /home/builder/.termux-build/nekovm/build/bin/neko: Exec format error
make[2]: *** [CMakeFiles/test.n.dir/build.make:76: bin/test.n] Error 2
/bin/sh: 1: /home/builder/.termux-build/nekovm/build/bin/neko: Exec format error
```

You have lots of packages in https://github.com/termux/termux-packages/tree/master/packages (including packages with hostbuilds) which could be used as base. Way to go.

**Comment by xeffyr** [comment]
Closing due to inactivity.
